packages/components/src/utils/hds-aria-described-by.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import { tracked } from '@glimmer/tracking';
+import { scheduleOnce } from '@ember/runloop';
+import type Component from '@glimmer/component';
+
+type ElementSet = Set<HTMLElement>;
+
+class AriaDescriptorMap {
+  private _elements = new WeakMap<AriaDescribedByComponent, ElementSet>();
+
+  register(obj: AriaDescribedByComponent, element: HTMLElement) {
+    const elements = this.findOrCreateElementSet(obj);
+    elements.add(element);
+  }
+
+  unregister(obj: AriaDescribedByComponent, element: HTMLElement) {
+    const elements = this.findOrCreateElementSet(obj);
+    elements.delete(element);
+  }
+
+  entries(obj: AriaDescribedByComponent) {
+    const elements = this.findOrCreateElementSet(obj);
+    return Array.from(elements.values());
+  }
+
+  private findOrCreateElementSet(obj: AriaDescribedByComponent): ElementSet {
+    if (this._elements.has(obj)) {
+      return this._elements.get(obj) as ElementSet;
+    }
+
+    const elements: ElementSet = new Set();
+    this._elements.set(obj, elements);
+    return elements;
+  }
+}
+
+const ariaDescriptorMap = new AriaDescriptorMap();
+
+interface AriaDescribedByArgs {
+  Args: {
+    extraAriaDescribedBy: string;
+  };
+}
+
+interface AriaDescribedByComponent extends Component<AriaDescribedByArgs> {
+  __ARIA_DESCRIPTION_IDS__: string[];
+  ariaDescribedBy: string;
+}
+
+// essentially a type that says we return a subclass of the given type T
+type ClassOf<T> = new () => T;
+
+export function ariaDescribedBy(
+  klass: ClassOf<AriaDescribedByComponent>
+): ClassOf<AriaDescribedByComponent> {
+  class X extends klass {
```

Comment:
trying to come up with more descriptive names here; what do you think about this?
```suggestion
export function ariaDescribedBy(
  BaseComponent: ClassOf<AriaDescribedByComponent>
): ClassOf<AriaDescribedByComponent> {
  class EnhancedComponent extends BaseComponent {
```